Add basic streaming controller

bhalevy opened the pull request:
streaming_controller adjust streaming shares based on
disk utilization, using the following configuration options:
- streaming_min_shares: minimum amount of shares for streaming.
- streaming_max_shares: maximum amount of shares for streaming.
- streaming_controller_disk_utilization_threshold: percentage of disk utilization at which
  the streaming controller starts adjusting straming shares.
    
When the node's disk utilization is less than `streaming_controller_disk_utilization_threshold`
streaming is given `streaming_min_shares`.  Otherwise, the number of shares set for streaming
is calculated using the following linear formula:
      factor = (disk_utilization - threshold) / (1 - threshold)
      streaming_min_shares + factor * (streaming_max_shares - streaming_min_shares)
    
Fixes #22804
Refs #23042

* Backport to 2025.1 is required for helping with out-of-space prevention during scale out at 90% utilization

bhalevy commented:
Cc @pehala @mykaul 

Petr, I'd like to give this branch a quick elasticity test at 90% utilization
with homogeneous and heterogeneous clusters (different instance type in the same instance family so to have different disk capacity per shard ratios).

The default configuration of {200, 1000, 0.9} *should* work out of the box (start with 200 shares, boost linearly up to 1000 above 90% utilization), but we might need to tweak it based on the test result.

In any case, please make sure `stream_io_throughput_mb_per_sec` is set to `0` (unlimited).
In the future the streaming_controller will adjust it too, if set, but at the moment it doesn't do it.
We will test and develop this feature once the basic controller gets in.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15536/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15536/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/streaming_controller_test
#### Build Details:
- Duration: 1 hr 37 min
- Builder: i-031f9b37aacff7080 (m5ad.12xlarge)

avikivity commented:
> streaming_controller adjust streaming shares based on disk utilization, using the following configuration options:
> 
> * streaming_min_shares: minimum amount of shares for streaming.
> * streaming_max_shares: maximum amount of shares for streaming.
> * streaming_controller_disk_utilization_threshold: percentage of disk utilization at which
>   the streaming controller starts adjusting straming shares.
> 
> When the node's disk utilization is less than `streaming_controller_disk_utilization_threshold` streaming is given `streaming_min_shares`. Otherwise, the number of shares set for streaming is calculated using the following linear formula: factor = (disk_utilization - threshold) / (1 - threshold) streaming_min_shares + factor * (streaming_max_shares - streaming_min_shares)
> 
> Fixes #22804
> 
> * Backport to 2025.1 is required for helping with out-of-space prevention during scale out at 90% utilization

I don't see any design or testing report.

bhalevy commented:
> > streaming_controller adjust streaming shares based on disk utilization, using the following configuration options:
> > 
> > * streaming_min_shares: minimum amount of shares for streaming.
> > * streaming_max_shares: maximum amount of shares for streaming.
> > * streaming_controller_disk_utilization_threshold: percentage of disk utilization at which
> >   the streaming controller starts adjusting straming shares.
> > 
> > When the node's disk utilization is less than `streaming_controller_disk_utilization_threshold` streaming is given `streaming_min_shares`. Otherwise, the number of shares set for streaming is calculated using the following linear formula: factor = (disk_utilization - threshold) / (1 - threshold) streaming_min_shares + factor * (streaming_max_shares - streaming_min_shares)
> > Fixes #22804
> > 
> > * Backport to 2025.1 is required for helping with out-of-space prevention during scale out at 90% utilization
> 
> I don't see any design or testing report.

Yes, at this stage, it's just a placeholder for the intention to backport.
I asked @pehala for testing [here](https://github.com/scylladb/scylladb/pull/23000#issuecomment-2677828496)
Design originated as part of https://docs.google.com/document/d/1wFvFO8t2vUBx5P-LhvA4bgT004URsIc-G-QyqWAbRXE/edit?tab=t.8ipvamac0mpl#heading=h.gd8e53ic23kt
I can separate it into a confluence page of its own.

bhalevy commented:
[530450b](https://github.com/scylladb/scylladb/commit/530450bfa81d36849457c262a9610ea10b04da19):
- Added `BOOST_AUTO_TEST_SUITE(streaming_controller_test)` to unit test

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15548/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15548/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/streaming_controller_test
#### Failed Tests (1/1):
- [test_streaming_controller_shares](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15548/testReport/junit/boost/streaming_controller_test/test_streaming_controller_shares) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_streaming_controller_shares)
#### Build Details:
- Duration: 1 hr 11 min
- Builder: i-030cef7a1a91f2510 (m5d.16xlarge)

bhalevy commented:
In d3c9dbc: REQUIRE_EVENTUALLY_EQUAL for live update to be propagated.
As seen in debug mode it may require context switch(es), so add a retry mechanism that would allow yielding.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15570/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15570/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/streaming_controller_test
#### Failed Tests (1/1):
- [test_streaming_controller_shares](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15570/testReport/junit/boost/streaming_controller_test/test_streaming_controller_shares) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_streaming_controller_shares)
#### Build Details:
- Duration: 48 min
- Builder: spider2.cloudius-systems.com

bhalevy commented:
In v2 ([d3c9dbc](https://github.com/scylladb/scylladb/commit/d3c9dbcc43efb3c358f2334ca296dd8a7f9bb1a8)):
- rebased
- improved controller log messages

bhalevy commented:
In v2.1 ([1d8a36d](https://github.com/scylladb/scylladb/commit/1d8a36d8d3e17dae4baa358034a9bf9ae8d28391)):
- deflaked unit test again for debug mode to tolerate cross shard inconsistency during shares adjustment.

bhalevy commented:
> > streaming_controller adjust streaming shares based on disk utilization, using the following configuration options:
> > 
> > * streaming_min_shares: minimum amount of shares for streaming.
> > * streaming_max_shares: maximum amount of shares for streaming.
> > * streaming_controller_disk_utilization_threshold: percentage of disk utilization at which
> >   the streaming controller starts adjusting straming shares.
> > 
> > When the node's disk utilization is less than `streaming_controller_disk_utilization_threshold` streaming is given `streaming_min_shares`. Otherwise, the number of shares set for streaming is calculated using the following linear formula: factor = (disk_utilization - threshold) / (1 - threshold) streaming_min_shares + factor * (streaming_max_shares - streaming_min_shares)
> > Fixes #22804
> > 
> > * Backport to 2025.1 is required for helping with out-of-space prevention during scale out at 90% utilization
> 
> I don't see any design or testing report.

See doc: https://scylladb.atlassian.net/wiki/spaces/RND/pages/34504729/Streaming+Controller+Requirement+Document

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15572/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15572/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/streaming_controller_test
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3648/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6766/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14790/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15572/)
#### Build Details:
- Duration: 5 hr 6 min
- Builder: spider4.cloudius-systems.com

bhalevy commented:
Looking at https://argus.scylladb.com/tests/scylla-cluster-tests/30d4ffc4-ff04-4c05-ab16-718c9638d066, the streaming controller is doing its job:
```
$ grep streaming_controller */messages.log
elasticity-test-ubuntu-db-node-30d4ffc4-4/messages.log:2025-02-26T20:01:40.651+00:00 elasticity-test-ubuntu-db-node-30d4ffc4-4     !INFO | scylla[5401]:  [shard 0:main] streaming_controller - Adjusting streaming shares to 202.58731 due to high disk utilization of 90.0%%
...
elasticity-test-ubuntu-db-node-30d4ffc4-6/messages.log:2025-02-27T03:27:06.173+00:00 elasticity-test-ubuntu-db-node-30d4ffc4-6     !INFO | scylla[5364]:  [shard 0:main] streaming_controller - Adjusting streaming shares to 556.1691 due to high disk utilization of 94.5%%
```

bhalevy commented:
In v3 (93a3fb0):
- rebased
- renamed threshold config option to a generic one: `high_disk_utilization_threshold`
- changed type for min/max streaming shares to unsigned
  - reduces spam in `adjust_if_needed` due to small fractional change in calculated shares.

bhalevy commented:
@paszkow I renamed the config option to a generic one.
Please add `critical_disk_utilization_level` for write rejection.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15667/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15667/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/streaming_controller_test
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14854/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3715/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6831/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15667/)
#### Build Details:
- Duration: 5 hr 3 min
- Builder: spider4.cloudius-systems.com

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15667/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15667/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/streaming_controller_test ✅ - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14854/) ❌ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3715/) ✅ - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6831/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15667/)

Unrelated: https://github.com/scylladb/scylladb/issues/22834

bhalevy commented:
Closing this as unplanned
